operator/validator/metadata/syncer.go

```diff
@@ -0,0 +1,366 @@
+	"go.uber.org/zap"
+
+	"github.com/ssvlabs/ssv/logging/fields"
+	networkcommons "github.com/ssvlabs/ssv/network/commons"
+	"github.com/ssvlabs/ssv/network/records"
+	"github.com/ssvlabs/ssv/protocol/v2/blockchain/beacon"
+	ssvtypes "github.com/ssvlabs/ssv/protocol/v2/types"
+	registrystorage "github.com/ssvlabs/ssv/registry/storage"
+	"github.com/ssvlabs/ssv/storage/basedb"
+)
+
+//go:generate mockgen -package=metadata -destination=./mocks.go -source=./syncer.go
+
+const (
+	defaultSyncInterval      = 12 * time.Minute
+	defaultStreamInterval    = 2 * time.Second
+	defaultUpdateSendTimeout = 30 * time.Second
+	batchSize                = 512
+)
+
+type Syncer struct {
+	logger            *zap.Logger
+	shareStorage      shareStorage
+	validatorStore    selfValidatorStore
+	beaconNetwork     beacon.BeaconNetwork
+	beaconNode        beacon.BeaconNode
+	syncInterval      time.Duration
+	streamInterval    time.Duration
+	updateSendTimeout time.Duration
+}
+
+type shareStorage interface {
+	List(txn basedb.Reader, filters ...registrystorage.SharesFilter) []*ssvtypes.SSVShare
+	Range(txn basedb.Reader, fn func(*ssvtypes.SSVShare) bool)
+	UpdateValidatorsMetadata(map[spectypes.ValidatorPK]*beacon.ValidatorMetadata) error
+}
+
+type selfValidatorStore interface {
+	SelfValidators() []*ssvtypes.SSVShare
+}
+
+func NewSyncer(
+	logger *zap.Logger,
+	shareStorage shareStorage,
+	validatorStore selfValidatorStore,
+	beaconNetwork beacon.BeaconNetwork,
+	beaconNode beacon.BeaconNode,
+	opts ...Option,
+) *Syncer {
+	u := &Syncer{
+		logger:            logger,
+		shareStorage:      shareStorage,
+		validatorStore:    validatorStore,
+		beaconNetwork:     beaconNetwork,
+		beaconNode:        beaconNode,
+		syncInterval:      defaultSyncInterval,
+		streamInterval:    defaultStreamInterval,
+		updateSendTimeout: defaultUpdateSendTimeout,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return u
+}
+
+type Option func(*Syncer)
+
+func WithSyncInterval(interval time.Duration) Option {
+	return func(u *Syncer) {
+		u.syncInterval = interval
+	}
+}
+
+func (s *Syncer) SyncOnStartup(ctx context.Context) (map[spectypes.ValidatorPK]*beacon.ValidatorMetadata, error) {
+	// Load non-liquidated shares.
+	shares := s.shareStorage.List(nil, registrystorage.ByNotLiquidated())
+	if len(shares) == 0 {
+		s.logger.Info("could not find non-liquidated validator shares on initial metadata retrieval")
+		return nil, nil
+	}
+
+	subnetsBuf := new(big.Int)
+	ownSubnets := s.selfSubnets(subnetsBuf)
+
+	// Skip syncing if metadata was already fetched before
+	// to prevent blocking startup after first sync.
+	needToSync := false
+	pubKeysToFetch := make([]spectypes.ValidatorPK, 0, len(shares))
+	for _, share := range shares {
+		networkcommons.SetCommitteeSubnet(subnetsBuf, share.CommitteeID())
+		subnet := subnetsBuf.Uint64()
+		if ownSubnets[subnet] == 0 {
+			continue
+		}
+		pubKeysToFetch = append(pubKeysToFetch, share.ValidatorPubKey)
```

Comment:
what do you think of filtering the shares by the subnets? I'm not sure we need to go over the shares twice..

Reply:
Yes, we could. We'd need to iterate once again to check if any of them doesn't have a metadata, although we'd be able to `break` the loop in this case

Reply:
sounds better to me, in current code you have to go over all the shares in the db twice.

Reply:
we can break it to a small task to the future though

Reply:
changed, but actually we still need to iterate all shares once again because we need a public key list, not a share list